### ISSUE TITLE
Naive chunks' cycle detect assertion 

### ISSUE THREAD

**Issue by ruiconti** [evanw/esbuild#3068]
## Context

I work on a custom inner loop bundle tool on top of esbuild, and I've noticed a regression on our build times. We work with the concept of "strategy", which is a way to coordinate more than one build. One of those is a "vendor strategy", which is analogous to vite's vendor build, which creates one build for user code, and another for vendor code. This build was taking a ridiculous amount of time (~470s on a sufficiently resourceful hardware, 64G 32vCPUs). 

## Investigation

After collecting some profiles, I've found that `linkerContext.enforceNoCyclicChunkImports` was behaving strangely:

![](https://user-images.githubusercontent.com/1834568/232511395-1f698e45-c565-4e1e-9604-86ed4817bd07.png)

Specially so when considering that its behavior is of an assert-like nature; that is, to ensure that the chunking algorithm behaved properly.

https://github.com/evanw/esbuild/blob/f0704baa38ef9a808b7e34cac75aa4ac285bb087/internal/linker/linker.go#L454-L478
 
It's a straightforward [cycle detection](https://en.wikipedia.org/wiki/Cycle_(graph_theory)#Cycle_detection) algorithm. However, given that our app is _huge_ and has large-enough dependency graph and connected-enough edge numbers, we have stress tested this algorithm and climbed the $N^2$ slope enough to be noticeable.

### But why so slow?

The worst-case time complexity of this algorithm can be broke down to:
- A DFS run: $O(N+E)$
- That is run for all chunks: $O(N * (N+E))$

where $N$ is the number of chunks, and $E$ is the number of edges on the graph.

In short, it's performing poorly because it is a brute force algorithm, that is being tested on a sufficiently large input space: $N > 2,500$. Breaking down what the algorithm does:

1. Iterates over all chunks and calls `validate(chunk)`
2. Compares the current chunk index with the indices in `path`, which is tracking the visited nodes.
3. Recursively calling validate for all `chunk.crossChunkImports`.

But because we are both iterating on all chunks _and_ recursively calling `validate` for all `chunk.crossChunkImports` without no mechanism to reduce the search space (i.e leave early), we are bound to do _lots_ of repetitive work. That is true to the amount of edges on the graph, because that would mean longer `validate` call stacks for each node that would be repeated over and over.
